### 1. Importação das bibliotecas essenciais

In [2]:
# Importando as bibliotecas
import pandas as pd
import plotly.express as px

print("Bibliotecas importadas com sucesso!")

Bibliotecas importadas com sucesso!


### 2. Carregamento dos dados

In [3]:
# Definindo o caminho do arquivo csv
caminho_arquivo = '../data/raw/base_churn_preditiva_20191030.csv'

# Lendo o arquivo, especificando que o separador é ';'
df = pd.read_csv(caminho_arquivo, sep=';')

print("As 5 primeiras linhas do DataFrame:")
display(df.head())

As 5 primeiras linhas do DataFrame:


ClientId DataExtracao  Score_Credito          Estado    Gênero  Idade  \
0   345.568   2019-06-30            619       São Paulo  Feminino     42   
1   345.569   2019-06-30            608  Rio de Janeiro  Feminino     41   
2   345.570   2019-06-30            502       São Paulo  Feminino     42   
3   345.571   2019-06-30            699       São Paulo  Feminino     39   
4   345.572   2019-06-30            850  Rio de Janeiro  Feminino     43   

   Tempo_Cliente Limite_Credito_Mercado  Qte_Categorias  Usa_Cartao_Credito  \
0              2                      0               1                   1   
1              1               838,0786               1                   0   
2              8              1.596,608               3                   1   
3              1                      0               2                   0   
4              2             1.255,1082               1                   1   

   Programa_Fidelidade Sum_Pedidos_Acumulados DataUltimaTransacao  Churn  
0                    1                422,287          2019-09-14      1  
1                    1         468,9274166667          2019-10-05      0  
2                    0             474,714875          2019-08-20      1  
3                    0         390,9442916667          2019-10-21      0  
4                    1         329,5170833333          2019-10-26      0

### 3. Diagnóstico inicial dos dados

In [6]:
# 1. Qual o tamanho do nosso dataset? (Linhas, Colunas)
print(f"O DataFrame possui {df.shape[0]} linhas e {df.shape[1]} colunas.")

O DataFrame possui 10000 linhas e 14 colunas.


In [7]:
# 2. Quais os tipos de dados e há valores faltando?
print("Tipos de Dados e Contagem de Nulos por Coluna:")
df.info()

Tipos de Dados e Contagem de Nulos por Coluna:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ClientId                10000 non-null  float64
 1   DataExtracao            10000 non-null  object 
 2   Score_Credito           10000 non-null  int64  
 3   Estado                  10000 non-null  object 
 4   Gênero                  10000 non-null  object 
 5   Idade                   10000 non-null  int64  
 6   Tempo_Cliente           10000 non-null  int64  
 7   Limite_Credito_Mercado  10000 non-null  object 
 8   Qte_Categorias          10000 non-null  int64  
 9   Usa_Cartao_Credito      10000 non-null  int64  
 10  Programa_Fidelidade     10000 non-null  int64  
 11  Sum_Pedidos_Acumulados  10000 non-null  object 
 12  DataUltimaTransacao     10000 non-null  object 
 13  Churn                   10000 non-null  int64

In [8]:
# 3. Como as variáveis numéricas se distribuem?
print("Estatísticas Descritivas das Variáveis Numéricas:")
display(df.describe())

Estatísticas Descritivas das Variáveis Numéricas:


ClientId  Score_Credito         Idade  Tempo_Cliente  \
count  10000.000000   10000.000000  10000.000000   10000.000000   
mean     350.567500     650.528800     38.921800       5.012800   
std        2.886896      96.653299     10.487806       2.892174   
min      345.568000     350.000000     18.000000       0.000000   
25%      348.067750     584.000000     32.000000       3.000000   
50%      350.567500     652.000000     37.000000       5.000000   
75%      353.067250     718.000000     44.000000       7.000000   
max      355.567000     850.000000     92.000000      10.000000   

       Qte_Categorias  Usa_Cartao_Credito  Programa_Fidelidade         Churn  
count    10000.000000         10000.00000         10000.000000  10000.000000  
mean         1.530200             0.70550             0.515100      0.201300  
std          0.581654             0.45584             0.499797      0.400992  
min          1.000000             0.00000             0.000000      0.000000  
25%          1.000000             0.00000             0.000000      0.000000  
50%          1.000000             1.00000             1.000000      0.000000  
75%          2.000000             1.00000             1.000000      0.000000  
max          4.000000             1.00000             1.000000      1.000000

In [10]:
# 4. Como as variáveis categóricas se distribuem?
# Selecionando apenas as colunas que são do tipo 'object' (texto)
df_categorico = df.select_dtypes(include="object")

# Aplicando o describe nesse novo DataFrame filtrado
print("Estatísticas Descritivas das Variáveis Categóricas:")
display(df_categorico.describe())

Estatísticas Descritivas das Variáveis Categóricas:


DataExtracao     Estado     Gênero Limite_Credito_Mercado  \
count         10000      10000      10000                  10000   
unique            1          3          2                   6382   
top      2019-06-30  São Paulo  Masculino                      0   
freq          10000       5014       5457                   3617   

       Sum_Pedidos_Acumulados DataUltimaTransacao  
count                   10000               10000  
unique                   9999                 122  
top            103,8538333333          2019-10-12  
freq                        2                 296

#### Análise dos resultados do Diagnóstico

- As colunas DataExtracao, Limite_Credito_Mercado, Sum_Pedidos_Acumulados e DataUltimaTransacao estão com o Dtype (tipo de dado) como object, ou seja, estão sendo tratadas como texto. Isso nos impede de fazer qualquer cálculo matemático ou análise de data com elas.

- Sem Dados Faltantes: a coluna Non-Null Count mostra 10000 para todas as variáveis. Não será preciso preencher dados ausentes.

- Dimensões: temos 10.000 clientes e 14 variáveis para analisar.

- Idade: a idade dos clientes varia de 18 a 92 anos, com uma média de aproximadamente 39 anos.

- Estado: temos clientes em 3 estados, sendo São Paulo o mais frequente.

- Churn: A taxa de Churn é de 20.13% (mean de 0.2013 na coluna Churn), o que é um número bem significativo.

### 4. Limpeza dos dados

In [11]:
# Lista das colunas numéricas que precisam de correção
cols_num_clean = ["Limite_Credito_Mercado", "Sum_Pedidos_Acumulados"]
for col in cols_num_clean:
    df[col] = df[col].astype(str).str.replace(".", "", regex=False)
    df[col] = df[col].str.replace(",", ".", regex=False).astype(float)

In [12]:
# Lista das colunas de data que precisam de correção
cols_date_clean = ["DataExtracao", "DataUltimaTransacao"]
for col in cols_date_clean:
    df[col] = pd.to_datetime(df[col])

In [15]:
print("\n--- Verificando os tipos de dados após a limpeza ---")
df.info()


--- Verificando os tipos de dados após a limpeza ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   ClientId                10000 non-null  float64       
 1   DataExtracao            10000 non-null  datetime64[ns]
 2   Score_Credito           10000 non-null  int64         
 3   Estado                  10000 non-null  object        
 4   Gênero                  10000 non-null  object        
 5   Idade                   10000 non-null  int64         
 6   Tempo_Cliente           10000 non-null  int64         
 7   Limite_Credito_Mercado  10000 non-null  float64       
 8   Qte_Categorias          10000 non-null  int64         
 9   Usa_Cartao_Credito      10000 non-null  int64         
 10  Programa_Fidelidade     10000 non-null  int64         
 11  Sum_Pedidos_Acumulados  10000 non-null  float64      